packages/worker/src/workflow-codec-runner.ts

```diff
@@ -0,0 +1,191 @@
+import { Payload, PayloadCodec, ProtoFailure } from '@temporalio/common';
+import { decodeFailure, encodeFailure } from '@temporalio/internal-non-workflow-common';
+import { coresdk } from '@temporalio/proto';
+import { clone, setWith } from 'lodash';
+
+/**
+ * Helper class for decoding Workflow activations and encoding Workflow completions.
+ */
+export class WorkflowCodecRunner {
+  constructor(protected readonly codec: PayloadCodec) {}
+
+  /**
+   * Run codec.decode on the Payloads in the Activation message.
+   */
+  public async decodeActivation(
+    activation: coresdk.workflow_activation.IWorkflowActivation
+  ): Promise<coresdk.workflow_activation.IWorkflowActivation> {
```

Comment:
This (and the similar stuff in the codec helpers file) seems maybe too verbose / change-sensitive to me.

Typically serialization concerns are handled with some sort of visitor pattern. IE: All fields of the struct are visited (unless explicitly excluded), and then de/serialized according to that field's type and possibly additional metadata. Here, we're visiting certain known fields and then handling them directly. It seems we could both shorten the code and make it more flexible in the face of change by visiting (transitively) any field with the `Payload` type, and applying the decode to it.

Of course that sort of type info is lost at runtime in TS so maybe we do what's described here https://stackoverflow.com/a/57594451/295987 and we just make a list of "here's all the payload fields we care about decoding" and that can be double-checked at compile time so if the interface changes we aren't caught by surprise.

Reply:
I considered visiting all fields and duck typing Payloads, but I guess there's a chance of having something else with `{ data, metadata }`, and doing specific fields is faster. I guess the speed difference is negligible and maintainability is more important.

Thanks for the link, will address in next PR.

Reply:
I suggested using the type checker to verify all fields are encoded.
I sent @lorensr some sample code that can handle that.
Can wait for the next PR.